unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp

```diff
@@ -0,0 +1,1446 @@
+        }
+        {
+          const java_generic_parametert &generic_param = generic_variables[1];
+          REQUIRE(is_java_generic_inst_parameter(generic_param));
+          REQUIRE(
+            generic_param ==
+            java_generic_inst_parametert(
+              symbol_typet(
+                "java::Interface_Implementation")));
+        }
+      }
+    }
+
+    THEN("The field3 component should be a pointer to "
+      "GenericClassWithGenericInnerClasses$GenericInnerClass")
+    {
+      const struct_typet::componentt &field_component =
+        require_type::require_component(class_type, "field3");
+
+      require_type::require_pointer(
+        field_component.type(),
+        symbol_typet(
+          class_prefix + "$GenericInnerClass"));
+
+      THEN("The pointer should be generic")
+      {
+        REQUIRE(is_java_generic_type(field_component.type()));
+        const auto &generic_variables =
+          to_java_generic_type(field_component.type()).generic_type_variables();
+        REQUIRE(generic_variables.size() == 2);
+        {
+          const java_generic_parametert &generic_param = generic_variables[0];
+          REQUIRE(is_java_generic_parameter(generic_param));
+          REQUIRE(
+            generic_param.type_variable() ==
+            symbol_typet(class_prefix + "::T"));
+        }
+        {
+          const java_generic_parametert &generic_param = generic_variables[1];
+          REQUIRE(is_java_generic_parameter(generic_param));
+          REQUIRE(
+            generic_param.type_variable() ==
+            symbol_typet(class_prefix + "::T"));
+        }
+      }
+    }
+
+    THEN("The field4 component should be a pointer to GenericClassWithGeneric"
+      "InnerClasses$GenericInnerClass$DoublyNestedInnerClass")
+    {
+      const struct_typet::componentt &field_component =
+        require_type::require_component(class_type, "field4");
+
+      require_type::require_pointer(
+        field_component.type(),
+        symbol_typet(
+          class_prefix + "$GenericInnerClass$DoublyNestedInnerClass"));
+
+      THEN("The pointer should be generic")
+      {
+        REQUIRE(is_java_generic_type(field_component.type()));
+        const auto &generic_variables =
+          to_java_generic_type(field_component.type()).generic_type_variables();
+        REQUIRE(generic_variables.size() == 2);
+        {
+          const java_generic_parametert &generic_param = generic_variables[0];
+          REQUIRE(is_java_generic_parameter(generic_param));
+          REQUIRE(
+            generic_param.type_variable() ==
+            symbol_typet(class_prefix + "::T"));
+        }
+        {
+          const java_generic_parametert &generic_param = generic_variables[1];
+          REQUIRE(is_java_generic_parameter(generic_param));
+          REQUIRE(
+            generic_param ==
+            java_generic_inst_parametert(
+              symbol_typet(
+                "java::Interface_Implementation")));
+        }
```

Comment:
Consider factoring the above two blocks into a helper function that you can call on `generic_variables[0]` and  `generic_variables[1]` ?